### ISSUE TITLE
Operation .Success never called when using `flatMap`: what am I doing wrong here?

### ISSUE THREAD

**Issue by iandundas** [ReactiveKit/ReactiveKit#48]
Hi! Would be very grateful if anyone could let me know what I'm doing wrong here:

[github demo project](https://github.com/iandundas/ReactiveKitOperationHelp/blob/master/FlatmapExample/AppDelegate.swift)

The below code is a simple Operation, which is called via a `flatMap`. The Operation simply calls `observer.success()`. I'd expect it to call `print("success")` - but this is never reached. If, however, I swap the line of code within the Operation to call `.next("..")`, this does reach the `observe` block.

Could anyone tell me why `case .success` doesn't get called in the` observe` block?

```swift
import ReactiveKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?

    var disposable: DisposableType! // just in case
    
    func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
        
        func successOperation() -> Operation<String, NSError>{
            return Operation{observer in
                
                // observer.next("this *does* reach the observe block")
                
                observer.success() // this never reaches the observe block
                
                return SimpleDisposable()
            }
        }
        
        let a = ActiveStream<String>()
        let b = a.flatMap(OperationFlatMapStrategy.Latest) { whatever -> Operation<String, NSError> in
            return successOperation()
        }
        
        disposable = b.observe { event in
            switch event{
            case .Success:
                print("success") // never called
                
            default:
                print("anything else")
            }
        }
        
        a.next("sending something")
        
        return true
    }
}
```

Note, if I just observe the Operation directly, it does work:

```swift
disposable = successOperation().observe { event in
    switch event{
    case .Success:
        print("success") // this IS called now.
        
    default:
        print("anything else")
    }
}
```

**Comment by iandundas** [comment]
Does `flatMap` swallow `.Success`? Is it supposed to? (My expectation would be no)

**Comment by srdanrasic** [comment]
ActiveStream has not concept of "completion", so we cannot know that ActiveStream has finished generating events. You would need to do something like:

```swift
a.next("sending something")
a.success()
```

But that is not supported yet. It requires non-trivial underlying changes. Fortunately - I'm working on that already :) You can expect next version with that supported sometime next week.

**Comment by iandundas** [comment]
> ActiveStream has not concept of "completion"

I don't understand - I'm not trying to complete the ActiveStream, but rather receive notifications of the completion of the Operation which we flatMap into it. The ActiveStream doesn't itself need to complete for the completion events of the flatMapped Operation (for example, a network request) to be important. 

For example, say we're starting a network request operation for each event coming from the ActiveStream - and we want to receive the `.success()` event which the Operation sends - is this not currently possible? 

Not sure with your reference to calling `.success()` on an ActiveStream - possibly the situation described above is an underlying limitation of ReactiveKit and the future work you're talking about will resolve it?

Cheers

**Comment by srdanrasic** [comment]
In general, streams can only send one terminal event (.Success or .Failure) and after it no additional events are to be sent. It cannot send, for example, two .Success events.

So, you are doing

```swift
a.next("bob")
```

but what if some time later you also did 

```swift
a.next("jane")
```

?

Should we pass .Success from _bob_ event (i.e. from its mapped operation), we would have to do either:
1. Complete the stream. In that case _jane_ event would never come through.
2. Send .Success but also propagate events from _jane_ event. In that case we would break the definition of a stream.

Given that, we are left with only one possible solution:
* Send .Success when both outer stream and all inner streams complete.

Hope this makes things clearer.

As for ReactiveKit - yes, this is limitation of basic Stream type (cannot complete) that I'm working to improve now.

**Comment by iandundas** [comment]
Ah - I figured it would be more like *1*, where flatMapped Operation couldn't send any more events after it completed - but then if the earlier ActiveStream sent another value, there would be a new Operation flatMapped, to which the observer would then be subscribed.

Perhaps I'm headed in the wrong direction. How could I currently achieve something like the following? 

A Stream sends values which each trigger the starting of an Operation, with an observer bound to the events of each Operation.

For example:

Stream sends `Void` (e.g. a Button Tap), each Void value is flatMapped to a network request Operation, the Operation begins, optionally sends `.Next`, sends `.Next`, then either sends `.Success()` or `.Failure()`

As it is at the moment, we'd only get the .Next() events, but never success or failure. Is there a way to trigger an Operation via a button tap, and still have access to Success and Failure events?

Thanks for your help :)

**Comment by srdanrasic** [comment]
That possible only if you expect the button to be tapped just once. If you can guaranty that, that you can do something like:

```swift
  let b = a.flatMap(OperationFlatMapStrategy.Latest) { _ in
    Operation<Int, NoError>.succeeded(with: 0)
  }.take(1).flatMap(OperationFlatMapStrategy.Latest) { _ in
    return successOperation()
  }
```

First we flatMap stream to a dummy operation that just sends zero on each event. We'll ignore that. Ideally we'd use Void instead of Int, but that's compiler limitation we have to live with until I make some improvements in next version.

Then we take(1) which just takes first event from the stream (first button tap) and completes.

Finally we apply your flatMap to, now, our Int Operation that completes after first event.

If the button is allowed to be tapped multiple times then it's in its nature never to complete and you should consider different approaches. If I'm understanding, you want to react only after all event from an operation are sent? Maybe you could use collect operator?

```swift
let b = a.flatMap(OperationFlatMapStrategy.Latest) { whatever -> Operation<String, NSError> in
  return successOperation().collect()
}
```

That will make `b.observeNext` fire next only when each operation completes with all events it has produced in an array.

**Comment by iandundas** [comment]
Cool - plenty to think about there. Will mull it over now! Thanks for your help man - appreciated.

**Comment by iandundas** [comment]
p.s. 

>  If I'm understanding correctly, you want to react only after all events from an operation are sent? Maybe you could use collect operator?

Actually no - it was more to get the Operation events as they happened in the terminating (correct term?) `observe` block.

**Comment by srdanrasic** [comment]
Streams are completable in v2. (Note that v2 is still in beta.)

**Comment by rsaarloos** [comment]
Hi @srdanrasic, not sure if this is fixed in v2, I'm experiencing the same issues.

When observing a single operation I get a event if a `.Completed`, `.Next` or `.Failure` event was called on the operation. This is as expected.

However when I flatmap another operation on this operation, then the second operation is correctly called by the first operation calling next. However if either the first or the second operations calls complete then this event is not seen when observing the entire operation.

A very simplistic view on this:

```
operationA.flatmap( return operationB).observe{ event in 

    switch (event):
     case .Next: // this is called when operation B calls .next()
     case .Failure: // this is called when either operation A or operation B calls failure
     case .Completed: // this is never called (not on operation A completion or operation B completion)
}
```

So I want to react on OperationA calling .Completed, .Next and .Failed and I want to react on operation B being .Completed or .Failed (.Next not that interested in). So I expect in the observe closure the .Completed case to be called when either Operation A or Operation B is completed.

What am I doing wrong here, or is this as expected?

**Comment by srdanrasic** [comment]
Hi @rsaarloos, expected behaviour is that `.Completed` is sent when both `operationA` and `operationB` complete.

**Comment by rsaarloos** [comment]
Hi @srdanrasic, ok great the only problem is that in my current code it seems that the observe on the operation does not get called when either OperationA or OperationB call `.completed()`, only the `.failure()` and `.next()` are catched by the observe

Is this an issue in ReactiveKit? If required I can create a sample project for this.

**Comment by srdanrasic** [comment]
@rsaarloos **Both** OperationA and OperationB have to complete, not just one of them. If that's not the case then we have a bug :) Could you check if it works when both complete?

**Comment by rsaarloos** [comment]
A sorry about that, I misunderstood, I see indeed when both OperationA and OperationB are completed the observe is called, so no bug there :)

However what should I do when I want a behaviour that when OperationA is completed (so I don't want to continue in my flow just stop there) I can react on this. Because if I want to get into the observe closure, this means I also need to complete OperationB but currently I do not have a direct reference to OperationB from inside OperationA as it is created in the scope of the `flatMap` on OperationA

Hope this makes sense.

**Comment by srdanrasic** [comment]
Yeah, it does. Hmm, what's the problem your trying to solve with that approach?

**Comment by rsaarloos** [comment]
@srdanrasic:  So lets say I have a web service. I allows us to create an object (POST operation). It has two success scenarios:

- http status 200: everything went fine and the obeject was created
- http status 201: the object was not yet created, because the webservice needs to have some kind of validations from you first and it sends back in the body a value that you need to use to create your validation

In the first scenario (200) we are basically done, we don't need to do anything else so the operation flow can stop. And we show to the user a success screen

In the second scenario (201) we need to ask the user some input. This input is then used to make another web request to validate the operation we wanted to do in the first place. After this web service is successful we again show the user a success screen. 

So I have three operations that I flatmap together:
- create object (web service request)
- ask for user input (a view controller that is pushed in view)
- validate object (web service request)

But I can already be done after the first operation and sometimes I need validation and then I'm done after the third operation. That's why I need this early-out
